File: pathplannerlib/src/main/java/com/pathplanner/lib/auto/SwerveAutoBuilder.java
package com.pathplanner.lib.auto;

import com.pathplanner.lib.PathPlannerTrajectory;
import com.pathplanner.lib.commands.PPSwerveControllerCommand;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.CommandBase;
import edu.wpi.first.wpilibj2.command.Subsystem;
import java.util.HashMap;
import java.util.function.Consumer;
import java.util.function.Supplier;

public class SwerveAutoBuilder extends BaseAutoBuilder {
  private final SwerveDriveKinematics kinematics;
  private final PIDConstants translationConstants;
  private final PIDConstants rotationConstants;
  private final Consumer<SwerveModuleState[]> outputModuleStates;
  private final Consumer<ChassisSpeeds> outputChassisSpeeds;
  private final Subsystem[] driveRequirements;

  private final boolean useKinematics;

  /**
   * Create an auto builder that will create command groups that will handle path following and
   * triggering events.
   *
   * <p>This auto builder will use PPSwerveControllerCommand to follow paths.
   *
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param resetPose A consumer that accepts a Pose2d to reset robot odometry. This will typically
   *     be called once at the beginning of an auto.
   * @param translationConstants PID Constants for the controller that will correct for translation
   *     error
   * @param rotationConstants PID Constants for the controller that will correct for rotation error
   * @param outputChassisSpeeds A function that takes the output ChassisSpeeds from path following
   *     commands
   * @param eventMap Map of event marker names to the commands that should run when reaching that
   *     marker.
   * @param driveRequirements The subsystems that the path following commands should require.
   *     Usually just a Drive subsystem.
   */
  public SwerveAutoBuilder(
      Supplier<Pose2d> poseSupplier,
      Consumer<Pose2d> resetPose,
      PIDConstants translationConstants,
      PIDConstants rotationConstants,
      Consumer<ChassisSpeeds> outputChassisSpeeds,
      HashMap<String, Command> eventMap,
      Subsystem... driveRequirements) {
    super(poseSupplier, resetPose, eventMap, DrivetrainType.HOLONOMIC);

    this.kinematics = null;
    this.translationConstants = translationConstants;
    this.rotationConstants = rotationConstants;
    this.outputModuleStates = null;
    this.outputChassisSpeeds = outputChassisSpeeds;
    this.driveRequirements = driveRequirements;

    this.useKinematics = false;
  }

  /**
   * Create an auto builder that will create command groups that will handle path following and
   * triggering events.
   *
   * <p>This auto builder will use PPSwerveControllerCommand to follow paths.
   *
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param resetPose A consumer that accepts a Pose2d to reset robot odometry. This will typically
   *     be called once at the beginning of an auto.
   * @param kinematics The kinematics for the robot drivetrain.
   * @param translationConstants PID Constants for the controller that will correct for translation
   *     error
   * @param rotationConstants PID Constants for the controller that will correct for rotation error
   * @param outputModuleStates A function that takes raw output module states from path following
   *     commands
   * @param eventMap Map of event marker names to the commands that should run when reaching that
   *     marker.
   * @param driveRequirements The subsystems that the path following commands should require.
   *     Usually just a Drive subsystem.
   */
  public SwerveAutoBuilder(
      Supplier<Pose2d> poseSupplier,
      Consumer<Pose2d> resetPose,
      SwerveDriveKinematics kinematics,
      PIDConstants translationConstants,
      PIDConstants rotationConstants,
      Consumer<SwerveModuleState[]> outputModuleStates,
      HashMap<String, Command> eventMap,
      Subsystem... driveRequirements) {
    super(poseSupplier, resetPose, eventMap, DrivetrainType.HOLONOMIC);

    this.kinematics = kinematics;
    this.translationConstants = translationConstants;
    this.rotationConstants = rotationConstants;
    this.outputModuleStates = outputModuleStates;
    this.outputChassisSpeeds = null;
    this.driveRequirements = driveRequirements;

    this.useKinematics = true;
  }

  @Override
  public CommandBase followPath(PathPlannerTrajectory trajectory) {
    if (useKinematics) {
      return new PPSwerveControllerCommand(
          trajectory,
          poseSupplier,
          kinematics,
          pidControllerFromConstants(translationConstants),
          pidControllerFromConstants(translationConstants),
          pidControllerFromConstants(rotationConstants),
          outputModuleStates,
          driveRequirements);
    } else {
      return new PPSwerveControllerCommand(
          trajectory,
          poseSupplier,
          pidControllerFromConstants(translationConstants),
          pidControllerFromConstants(translationConstants),
          pidControllerFromConstants(rotationConstants),
          outputChassisSpeeds,
          driveRequirements);
    }
  }
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/auto/SwerveAutoBuilder.cpp
#include "pathplanner/lib/auto/SwerveAutoBuilder.h"

#include "pathplanner/lib/commands/PPSwerveControllerCommand.h"

using namespace pathplanner;

SwerveAutoBuilder::SwerveAutoBuilder(std::function<frc::Pose2d()> pose,
		std::function<void(frc::Pose2d)> resetPose,
		PIDConstants translationConstants, PIDConstants rotationConstants,
		std::function<void(frc::ChassisSpeeds)> output,
		std::unordered_map<std::string, std::shared_ptr<frc2::Command>> eventMap,
		std::initializer_list<frc2::Subsystem*> driveRequirements,
		bool useAllianceColor) : BaseAutoBuilder(pose, resetPose, eventMap,
		BaseAutoBuilder::DriveTrainType::HOLONOMIC, useAllianceColor), m_kinematics(
		frc::Translation2d(), frc::Translation2d(), frc::Translation2d(),
		frc::Translation2d()), m_translationConstants(translationConstants), m_rotationConstants(
		rotationConstants), m_outputSpeeds(output), m_driveRequirements(
		driveRequirements), m_useKinematics(false) {

}

SwerveAutoBuilder::SwerveAutoBuilder(std::function<frc::Pose2d()> pose,
		std::function<void(frc::Pose2d)> resetPose,
		frc::SwerveDriveKinematics<4> kinematics,
		PIDConstants translationConstants, PIDConstants rotationConstants,
		std::function<void(std::array<frc::SwerveModuleState, 4>)> output,
		std::unordered_map<std::string, std::shared_ptr<frc2::Command>> eventMap,
		std::initializer_list<frc2::Subsystem*> driveRequirements,
		bool useAllianceColor) : BaseAutoBuilder(pose, resetPose, eventMap,
		BaseAutoBuilder::DriveTrainType::HOLONOMIC, useAllianceColor), m_kinematics(
		kinematics), m_translationConstants(translationConstants), m_rotationConstants(
		rotationConstants), m_outputStates(output), m_driveRequirements(
		driveRequirements), m_useKinematics(true) {

}

frc2::CommandPtr SwerveAutoBuilder::followPath(
		PathPlannerTrajectory trajectory) {
	if (m_useKinematics) {
		return PPSwerveControllerCommand(trajectory, m_pose, m_kinematics,
				BaseAutoBuilder::pidControllerFromConstants(
						m_translationConstants),
				BaseAutoBuilder::pidControllerFromConstants(
						m_translationConstants),
				BaseAutoBuilder::pidControllerFromConstants(
						m_rotationConstants), m_outputStates,
				m_driveRequirements, m_useAllianceColor).ToPtr();
	} else {
		return PPSwerveControllerCommand(trajectory, m_pose,
				BaseAutoBuilder::pidControllerFromConstants(
						m_translationConstants),
				BaseAutoBuilder::pidControllerFromConstants(
						m_translationConstants),
				BaseAutoBuilder::pidControllerFromConstants(
						m_rotationConstants), m_outputSpeeds,
				m_driveRequirements, m_useAllianceColor).ToPtr();
	}
}
